netweaver/monitoring.sls

```diff
@@ -6,33 +6,34 @@ prometheus_sap_host_exporter_pkg:
     - name: prometheus-sap_host_exporter
 
 # the sid, instance number pair of a node is unique, so we need to adapt configuration
-{% for node in netweaver.nodes if host == node.host and node.sap_instance != "db" %}
-{% set instance = '{:0>2}'.format(node.instance) %}
-{% set daemon_instance = '{}_{}'.format(node.sid, instance) %}
+{% for node in netweaver.nodes if node.sap_instance != "db" %}
+{% set sap_instance_nr = '{:0>2}'.format(node.instance) %}
+{% set exporter_instance = '{}_{}{}'.format(node.sid, node.sap_instance.upper(), sap_instance_nr) %}
 
-sap_host_exporter_service_{{ daemon_instance }}:
-  service.running:
-    - name: prometheus-sap_host_exporter@{{ daemon_instance }}
-    - enable: True
-    - restart: True
-    - require:
-      - pkg: prometheus_sap_host_exporter_pkg
-      - file: sap_host_exporter_configuration_{{ daemon_instance }}
-    - watch:
-      - file: sap_host_exporter_configuration_{{ daemon_instance }}
-
-
-# we can have multiple exporters in the same host in that moment, we need to change the port used by the exporter
-sap_host_exporter_configuration_{{ daemon_instance }}:
+# we bind each exporter instance to a SAP instance virtual host
+sap_host_exporter_configuration_{{ exporter_instance }}:
   file.managed:
-    - name: /etc/sap_host_exporter/{{ daemon_instance }}.yaml
+    - name: /etc/sap_host_exporter/{{ exporter_instance }}.yaml
     - contents: |
-         port: "968{{ loop.index0 }}"
-         address: "0.0.0.0"
+         port: "9680"
+         address: {{ node.virtual_host }}
```

Comment:
Small consideration with this. Attaching the virtual host to the exporter will make the endpoint accessible only for this network, so only in this network. We are not going to be able to access the metrics from our computers.
For real usage is fine, but for development is annoying.
We cannot really do anything about this, but to keep in mind the restriction.

Reply:
well, there are a few ways around this:
- you can log in the node and browse the metrics locally with curl
- you can use the Prometheus dashboard

it's a small tradeoff to avoid some quirky port allocation logic within the salt state.